src/gluonts/mx/trainer/_base.py

```diff
@@ -204,250 +244,230 @@ def __call__(
-                net=net,
-                hybridize=self.hybridize,
-                static_alloc=True,
-                static_shape=True,
-            ):
-                best_epoch_info = {
-                    "params_path": "%s-%s.params" % (base_path(), "init"),
-                    "epoch_no": -1,
-                    "score": np.Inf,
-                }
+            optimizer = mx.optimizer.Adam(
+                learning_rate=self.learning_rate,
+                wd=self.weight_decay,
+                clip_gradient=self.clip_gradient,
+            )
 
-                lr_scheduler = lrs.MetricAttentiveScheduler(
-                    objective="min",
-                    patience=self.patience,
-                    decay_factor=self.learning_rate_decay_factor,
-                    min_lr=self.minimum_learning_rate,
-                )
+            trainer = mx.gluon.Trainer(
+                net.collect_params(),
+                optimizer=optimizer,
+                kvstore="device",  # FIXME: initialize properly
+            )
 
-                optimizer = mx.optimizer.Adam(
-                    learning_rate=self.learning_rate,
-                    lr_scheduler=lr_scheduler,
-                    wd=self.weight_decay,
-                    clip_gradient=self.clip_gradient,
-                )
+            first_forward = True
 
-                trainer = mx.gluon.Trainer(
-                    net.collect_params(),
-                    optimizer=optimizer,
-                    kvstore="device",  # FIXME: initialize properly
-                )
+            def loop(  # todo call run epoch
+                epoch_no,
+                batch_iter,
+                num_batches_to_use: Optional[int] = None,
+                is_training: bool = True,
+            ) -> mx.metric.Loss:
+                nonlocal first_forward
+                tic = time.time()
 
-                first_forward = True
+                epoch_loss = mx.metric.Loss()
 
-                def loop(
-                    epoch_no,
-                    batch_iter,
-                    num_batches_to_use: Optional[int] = None,
-                    is_training: bool = True,
-                ) -> mx.metric.Loss:
-                    nonlocal first_forward
-                    tic = time.time()
-
-                    epoch_loss = mx.metric.Loss()
-
-                    # use averaged model for validation
-                    if not is_training and isinstance(
-                        self.avg_strategy, IterationAveragingStrategy
-                    ):
-                        self.avg_strategy.load_averaged_model(net)
-
-                    batch_iter = itertools.islice(
-                        batch_iter, num_batches_to_use
+                if is_training:
+                    self.callbacks.on_train_batch_start(training_network=net)
+                else:
+                    self.callbacks.on_validation_batch_start(
+                        training_network=net
```

Comment:
@borchero yes, I think this should go at the beginning of the loop that follows (I may have screwed up the rebasing here)

Reply:
Should we keep the API as is and put this in the loop or use `on_train_epoch_start` and `on_validation_epoch_start` and keep the code where it is (i.e. before the loop).

`on_train_batch_start` is basically equivalent to `on_train_batch_end` for all but the first batch within an epoch (i.e. nothing changed between the first call to `on_train_batch_end` and the second call to `on_train_batch_start`).

Reply:
ah no you're right, so maybe this was on purpose -- I take back my previous comment